doc/specs/thrift-binary-protocol-encoding.md

```diff
@@ -0,0 +1,467 @@
+rejected.
+
+```
+Compact protocol (4+ bytes):
++--------+--------+--------+...+--------+--------+...+--------+--------+...+--------+
+|pppppppp|mmmvvvvv| seq id              | name length         | name                |
++--------+--------+--------+...+--------+--------+...+--------+--------+...+--------+
+```
+
+Where:
+
+* `pppppppp` is the protocol id, fixed to `1000 0010`, 0x82
+* `vvvvv` is the version, an unsigned 5 bit integer, fixed to `00001`, 0x1
+* `mmm` is the message type, an unsigned 3 bit integer
+* `seq id` is the sequence id, a signed 32 bit integer encoded as a var int.
+* `name length` is the byte length of the name field, a signed 32 bit integer encoded as a var int (must be >= 0).
+* `name` is the method name to invoke, a UTF-8 encoded string.
+
+A server could automatically determine whether a client talks the binary protocol or the compact protocol by
+investigating the first byte. If the value is `1000 0001` or `0000 0000` (assuming a name shorter then ±16 MB) it is the
+binary protocol. When the value is `1000 0010` it is talking the compact protocol.
+
+Message types are encoded with the following values:
+
+* _Call_: 1
+* _Reply_: 2
+* _Exception_: 3
+* _Oneway_: 4
+
+For a method name of 32 bytes, the binary protocol (strict) needs 44 bytes and the compact protocol needs 36 to 40 bytes.
+
+## Method arguments, return types and exceptions
+
+TODO: method arguments are encoded as a struct
+
+TODO: return value are encoded as a ?
+
+Both the binary protocol and compact protocol encode the result the same. The result is encoded as a struct with a
+single field. The field-id of that field is `0`. The type of the field corresponds to the type of the service method
+return type. When the return value is `null`, an empty struct is encoded.
+
+TODO: describe TupleProtocol also?
+
+TODO: exceptions are encoded as a struct with what field?
+
+### Structs and Unions
+
+Both the binary protocol and the compact protocol encode structs as a sequence of fields, followed by a stop field. Each
+field starts with a field header and is followed by the encoded field value. The encoding follows this BNF (`*` means 0
+or more times, parenthesis are used for grouping):
+
+```
+struct        => ( field-header field-value )* stop-field
+field-header  => field-type field-id
+```
+
+Because each field header contains the field-id (as defined by the IDL), the fields can be encoded in any order.
+Thrift's type system is not extensible, you can only encode the primitive types and structs. Therefore is also possible
+to handle unknown fields while decoding; these are simply ignored. While decoding the field type can be used to
+determine how to decode the field value.
+
+In the binary protocol the field header is always 3 bytes long. In the compact protocol the field header packs a lot
+more cleverness. In most cases it is 1 byte long. In special cases it can grow to 2 or 3 bytes. When your field-ids are
+very large it can even grow to 4 bytes.
```

Comment:
Field headers are more clever then that: they have a short and long form. In the short form they contain a field-id delta.